### PR TITLE
[FLINK-23617][datastream] Co-locate sink operators with same parallelism

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

FLINK-19531 introduced the implementation of the Sink interface. It strictly cut the different parts of the sink pipeline into 3 operators:

writer -> committer -> global committer

In streaming mode with a parallelism p, the pipeline is executed as follows
writer(parallelism=p) -> committer(parallelism=p) -> global committer(parallelism=1).
Here we could bundle writer+committer into one operator.

In batch mode with a parallelism p, the pipeline is executed as follows
writer(parallelism=p) -> committer(parallelism=1) -> global committer(parallelism=1).
Here we could bundle committer+global committer into one operator. (Committer needs to run with parallelism=1 to create a pipeline region and reduce the risk of dataloss during commit; we can hopefully fix it after FLIP-147)

Having fewer operators will decrease the need to copy the committables in the operator chain (where we currently mostly use Kryo). Thus, we can implement connection/transaction pooling in streaming, where committables are reused after successful commit.

The proposal of this ticket is to extract the functionality of the 7 different operator implementations with their factories into reusable building blocks and use them in 2 operators (writer+committer and committer+global committer).

## Brief change log

- Added 2 interfaces for managing writer state and handling committables
- Implemented building blocks with these two interfaces for streaming/batch (global) committers
- Added 2 operators that use the building blocks to provide writer+committer and committer+global committer
- Use the provided serializers to avoid using Kryo for committables by exchanges committables as `byte[]`

## Verifying this change

Adjusted the existing unit tests. FileSink ITCases also cover this refactoring.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+ * An operator that processes records to be written into a {@link
+ * org.apache.flink.api.connector.sink.Sink}. It also has a way to process committables with the
+ * same parallelism or send them downstream to a {@link CommitterOperator} with a different
+ * parallelism.
+ *
+ * <p>The operator may be part of a sink pipeline and is the first operator. There are currently two
+ * ways this operator is used:
+ *
+ * <ul>
+ *   <li>In streaming mode, there is a this operator with parallelism p containing {@link
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java:54]: Remove `a`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+/**
+ * A {@link org.apache.flink.streaming.api.operators.StreamOperatorFactory} for {@link
+ * SinkOperator}.
+ *
+ * @param <InputT> The input type of the {@link SinkWriter}.
+ * @param <CommT> The committable type of the {@link SinkWriter}.
+ * @param <WriterStateT> The type of the {@link SinkWriter Writer's} state.
+ */
+public final class SinkOperatorFactory<InputT, CommT, WriterStateT>
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java:40]: I wonder if this class should be part of the previous commit IMO it belongs to the newly introduced operator model.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java [hunk 3]
```diff
@@ -33,44 +30,52 @@
...
 
     /** Responsible for committing the committable to the external system. */
-    private final Committer<CommT> committer;
+    private final Committer<InputT> committer;
 
-    /** Record all the committables until the end of the input. */
-    private final List<CommT> allCommittables;
+    /**
+     * The committer that is chained to this committer. It's either {@link
+     * GlobalBatchCommitterHandler} or {@link NoopCommitterHandler}.
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java:43]: Wouldn't is be easier to just not instantiate a `GlobalBatchCommitter` if not necessary and not introduce the `NoopCommitterHandler`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            CommitterHandler<CommT, GlobalCommT> committerHandler = getGlobalCommitterHandler();
+            if (batch) {
+                Optional<Committer<CommT>> committer = sink.createCommitter();
+                if (committer.isPresent()) {
+                    committerHandler =
+                            new BatchCommitterHandler<>(committer.get(), committerHandler);
+                }
+            }
+
+            checkState(
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java:71]: Isn't this always failing for batch mode because the batch mode set the committerHandler to `BatchCommitterHandler`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    !(committerHandler instanceof NoopCommitterHandler),
+                    "committer operator without commmitter");
+            final CommitterOperator<CommT, GlobalCommT> committerOperator =
+                    new CommitterOperator<>(committableSerializer, committerHandler, null);
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java:75]: I prefer having a separate constructor when fields are nullable.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/GlobalBatchCommitterHandler.java [hunk 6]
```diff
@@ -72,11 +58,12 @@ public void endInput() throws Exception {
...
         }
         globalCommitter.endOfInput();
+        return Collections.emptyList();
     }
 
     @Override
     public void close() throws Exception {
-        super.close();
         globalCommitter.close();
+        super.close();
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/GlobalBatchCommitterHandler.java:67]: Why do we need this change? I am not fully familiar with the exact semantics but first closing the object itself and next the `globalCommitter` looks safer.

**Review by fapaul — COMMENTED** [review]
Thanks for the effort! I did a first pass and left a couple of questions but in general, the idea of combining the operators looks good.

I think we should also adapt the docstrings in the `Sink.java` to reflect the new invariants when user needs to provide Committable serializer

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java:40]: I agree but this class depends on the specific `CommitterHandler`s which I wanted to convert in this commit.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java:43]: The `NoopCommitterHandler` avoids null checks inside this class. It's a matter of flavor but in general a `NoopCommitterHandler` is the proper OOP approach (also dubbed case-less programming because of VMT).

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/FileSink.java [hunk 7]
```diff
@@ -161,6 +163,12 @@ private FileSink(BucketsBuilder<IN, ? extends BucketsBuilder<IN, ?>> bucketsBuil
         return Optional.empty();
     }
 
+    @Override
+    public Collection<String> getCompatibleStateNames() {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/FileSink.java:167]: I don't understand your comment. Why would you favor a `List` here? The only difference between a collection and a list is that the latter is implicitly ordered and indexed. None of that matters here.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java [hunk 8]
```diff
@@ -81,9 +83,17 @@
     /** Return the serializer of the writer's state type. */
     Optional<SimpleVersionedSerializer<WriterStateT>> getWriterStateSerializer();
 
+    /**
+     * A list of state names of sinks from which the state can be restored. For example, the new
+     * file sink can resume from the state of an old streaming file sink as a drop-in replacement
+     * when resuming from a checkpoint/savepoint.
+     */
+    default Collection<String> getCompatibleStateNames() {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java:91]: I consider it more of a bugfix ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    default void close() throws Exception {}
+
+    /** Called when a checkpoint is completed and returns a list of output to be sent downstream. */
+    default Collection<OutputT> notifyCheckpointCompleted(long checkpointId) throws IOException {
+        return Collections.emptyList();
+    }
+
+    /** Swallows all committables and emits nothing. */
+    @SuppressWarnings("unchecked")
+    static <InputT, OutputT> CommitterHandler<InputT, OutputT> noop() {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:74]: I'll remove it. Originally, `CommitterHandler` was public and the composition happened in the SinkTranslater.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java (hunk 4 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java:71]: It's always succeeding (`Precondition#checkState` fails on `false`) in your case. However, it could fail if there is no global committer in streaming and it fails if there is neither a committer nor a global committer in batch. But this is a programmatic error in `SinkTranslater`; hence the `checkState`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/GlobalBatchCommitterHandler.java (hunk 6 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/GlobalBatchCommitterHandler.java:67]: You are supposed to close resources always in the opposite order of creation. That's also how try-with-resource and Guava's Closer work. The main idea is that your later resources may depend on the prior resources to finish properly.
Here, it doesn't matter, but I'd still like to have it consistently correct in Flink.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java (hunk 3 above)
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java:43]: I do not necessarily see the relation to OOP here ;) but isn't it also some kind of optimization to not instantiate this Noop object?

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java:43]: Here is some reference https://en.wikipedia.org/wiki/Null_object_pattern TL;DR `null` is an external construct in OOP.

Also don't worry about performance here: First, we have a singleton pattern as an enum that's very efficient. Second, we are not on the hot path.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:74]: I can't fully remove it without introducing some ugliness: This method is for casting the singleton into the expected type. The `NoopCommitterHandler.INSTANCE` could actually be covariant to any type but you can't express that in Java. 
I have moved it to `NoopCommitterHandler#getInstance()`. PTAL if it's less confusing. I'm also promoting `NoopCommitterHandler` to top-level.

**Comment by AHeise** [comment]
> I think we should also adapt the docstrings in the `Sink.java` to reflect the new invariants when user needs to provide Committable serializer
I have improved the doc and would merge this new commit with the first commit later. Could you please check & improve?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ * @param <InputT>
+ * @param <OutputT>
+ */
+interface CommitterHandler<InputT, OutputT> extends AutoCloseable, Serializable {
+
+    /** Initializes the state of the committer and this handler. */
+    default void initializeState(StateInitializationContext context) throws Exception {}
+
+    /** Snapshots the state of the committer and this handler. */
+    default void snapshotState(StateSnapshotContext context) throws Exception {}
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:43]: We can also move these to methods to `AbstractStreamingCommitterHandler` because AFAIS they are not necessary for the batch committers.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     * Called when no more committables are going to be added through {@link
+     * #processCommittables(SupplierWithException)}.
+     *
+     * @return a list of output committables that is send downstream.
+     */
+    default List<OutputT> endOfInput() throws IOException {
+        return Collections.emptyList();
+    }
+
+    default void close() throws Exception {}
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:65]: Nit: If the default implementation is only for `NoopCommitterHandler` we can think of just implementing it there.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            SupplierWithException<List<InputT>, IOException> committableSupplier)
+            throws IOException;
+
+    /**
+     * Called when no more committables are going to be added through {@link
+     * #processCommittables(SupplierWithException)}.
+     *
+     * @return a list of output committables that is send downstream.
+     */
+    default List<OutputT> endOfInput() throws IOException {
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:61]: Out of curiosity why does the `StreamingCommitterHandler` not implement `endOfInput`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.flink.util.function.SupplierWithException;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A wrapper around multiple {@link org.apache.flink.api.connector.sink.Committer} or {@link
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:31]: Nit: I find the word `multiple` a bit confusing here because afait it always exactly wraps one instance of a committer.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/TestHarnessUtil.java [hunk 14]
```diff
@@ -139,6 +139,7 @@ public static void assertNoLateRecords(Iterable<Object> elements) {
 
         testHarness.processElements(
                 input.stream().map(StreamRecord::new).collect(Collectors.toList()));
+        testHarness.prepareSnapshotPreBarrier(1);
```
Comment (fapaul) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/TestHarnessUtil.java:142]: Nit: Why was this changed needed?

**Review by fapaul — COMMENTED** [review]
Thanks for addressing most of my comments and the helpful explanations. I left some last nitty things but overall is looks already pretty good.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java (hunk 12 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:61]: Good questions. I double-checked and EOI on streaming is really just meant for flushing data. A final checkpoint happens afterwards (especially after FLIP-147).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/TestHarnessUtil.java (hunk 14 above)
Comment (AHeise) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/util/TestHarnessUtil.java:142]: Since writer+committer are merged, we now need to simulate more of the actual calls. The writer creates the committables on `prepareSnapshotPreBarrier` and pass it to the committer which then does it's actual work on `snapshotState`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java (hunk 10 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:43]: They are called for all `CommitterHandler` though so we need to leave them here. Or we have to do instanceof in the operators.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java [hunk 15]
```diff
@@ -46,44 +54,81 @@
...
+     * Returns the serializer of the aggregated committable type. The serializer is required iff the
+     * sink has a {@link GlobalCommitter}.
+     */
     Optional<SimpleVersionedSerializer<GlobalCommT>> getGlobalCommittableSerializer();
 
-    /** Return the serializer of the writer's state type. */
-    Optional<SimpleVersionedSerializer<WriterStateT>> getWriterStateSerializer();
+    /**
+     * A list of state names of sinks from which the state can be restored. For example, the new
+     * file sink can resume from the state of an old streaming file sink as a drop-in replacement
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java:123]: I can't exactly use a JavaDoc link because of circular dependency issues. But I have used the proper class names in the code tag to make it more obvious.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java [hunk 16]
```diff
@@ -48,41 +54,70 @@
...
+     * @throws IOException for any failure during creation.
      */
     Optional<Committer<CommT>> createCommitter() throws IOException;
 
     /**
-     * Creates a {@link GlobalCommitter}.
+     * Creates a {@link GlobalCommitter} which is part of a 2-phase-commit protocol. The {@link
+     * SinkWriter} creates committables through {@link SinkWriter#prepareCommit(boolean)} in the
+     * first phase. The committables are then passed to the Committer and persisted with {@link
+     * Committer#commit(List)} which also can return an aggregated committable. This aggregated
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java:99]: You are right. Thanks for double-checking!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java (hunk 10 above)
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterHandler.java:43]: I am just a bit unhappy with the overall structure of having this kind of deep inheritance but I guess refactoring this is probably out of scope.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java [hunk 17]
```diff
@@ -49,10 +49,10 @@
  * @param <InputT> The input type of the {@link Committer}.
  * @param <CommT> The committable type of the {@link Committer}.
  */
-abstract class AbstractStreamingCommitterOperator<InputT, CommT>
-        extends AbstractStreamOperator<CommT> implements OneInputStreamOperator<InputT, CommT> {
-
-    private static final long serialVersionUID = 1L;
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java:55]: Is it safe to remove this line?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A {@link org.apache.flink.streaming.api.operators.StreamOperatorFactory} for {@link
+ * CommitterOperator}.
+ *
+ * @param <CommT> the type of the committable
+ * @param <GlobalCommT> the type of the global committable
+ */
+public final class CommitterOperatorFactory<CommT, GlobalCommT>
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java:42]: Add `@Internal` annotation?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java [hunk 19]
```diff
@@ -32,25 +34,42 @@
...
+ * @param <InputT> the type of the committable
+ * @param <CommT> the type of the committable (to send to downstream operators)
+ * @param <WriterStateT> the type of the writer state for stateful sinks
  */
-@Internal
-abstract class AbstractSinkWriterOperator<InputT, CommT> extends AbstractStreamOperator<CommT>
-        implements OneInputStreamOperator<InputT, CommT>, BoundedOneInput {
-
-    private static final long serialVersionUID = 1L;
+public class SinkOperator<InputT, CommT, WriterStateT> extends AbstractStreamOperator<byte[]>
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java:71]: `@Internal`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.Optional;
+
+/**
+ * A {@link org.apache.flink.streaming.api.operators.StreamOperatorFactory} for {@link
+ * SinkOperator}.
+ *
+ * @param <InputT> The input type of the {@link SinkWriter}.
+ * @param <CommT> The committable type of the {@link SinkWriter}.
+ * @param <WriterStateT> The type of the {@link SinkWriter Writer's} state.
+ */
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperatorFactory.java:40]: `@Internal`

**Review by fapaul — APPROVED** [review]
I left a couple of last nitty comments but overall it looks ready to go :shipit:

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java (hunk 17 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java:55]: You mean `serialVersionUID`? With `StreamOperatorFactory` (which are serializable) there is actually no need for operators to be serializable anymore and it will be removed in the long run.

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java:55]: Note that factories should have `serialVersionUID` if we are very strict. However, actually in non-persistent serialization none of it actually matters. `serialVersionUID` is only important if you need a way to read old serialized data into changed serializable classes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java (hunk 19 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkOperator.java:71]: I made it non-public instead.
